{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f03924bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, SimpleLocation, CompoundLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bb679fee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b498615a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create features\n",
    "data_path = Path(\"/mnt/d/data/popnet_paper/variant_data/\")\n",
    "plasmo_gene_data = pd.read_csv(data_path / 'plasmo_genes.csv')\n",
    "neis_gene_data = pd.read_csv(data_path / 'neis_genes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "641c51be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plamsmo stuff\n",
    "gene_features = []\n",
    "for gene, rows in plasmo_gene_data.groupby('name'):\n",
    "    # cds features\n",
    "    is_forward = rows['direction'].unique()[0] == 'forward'\n",
    "    qualifiers = {\n",
    "        'segment': int(rows.region.unique()[0].split(':')[1]),\n",
    "        'segment_length': 5000,\n",
    "    }\n",
    "    if len(rows) > 1:\n",
    "        gene_features.append(SeqFeature(\n",
    "            CompoundLocation(\n",
    "                [\n",
    "                    SimpleLocation(\n",
    "                        r['cds_start'], r['cds_end'],\n",
    "                        strand = 1 if is_forward else -1,\n",
    "                    ) for r in rows.to_dict('records')\n",
    "                ]\n",
    "            ),\n",
    "            type = 'gene', id = gene, qualifiers = qualifiers,\n",
    "        ))\n",
    "    else:\n",
    "        row = rows.iloc[0]\n",
    "        gene_features.append(SeqFeature(SimpleLocation(\n",
    "            int(row.cds_start), int(row.cds_end),\n",
    "            strand = 1 if row.direction == 'forward' else -1\n",
    "        ), type = 'gene', id = gene, qualifiers = qualifiers,))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7f1f6e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene, rows in neis_gene_data.groupby('name'):\n",
    "    # cds features\n",
    "    is_forward = rows['direction'].unique()[0] == 'forward'\n",
    "    qualifiers = {\n",
    "        'segment': int(rows.region.unique()[0]),\n",
    "        'segment_length': 1000,\n",
    "    }\n",
    "    if len(rows) > 1:\n",
    "        gene_features.append(SeqFeature(\n",
    "            CompoundLocation(\n",
    "                [\n",
    "                    SimpleLocation(\n",
    "                        r['cds_start'], r['cds_end'],\n",
    "                        strand = 1 if is_forward else -1,\n",
    "                    ) for r in rows.to_dict('records')\n",
    "                ]\n",
    "            ),\n",
    "            type = 'gene', id = gene, qualifiers = qualifiers,\n",
    "        ))\n",
    "    else:\n",
    "        row = rows.iloc[0]\n",
    "        gene_features.append(SeqFeature(SimpleLocation(\n",
    "            int(row.cds_start), int(row.cds_end),\n",
    "            strand = 1 if row.direction == 'forward' else -1\n",
    "        ), type = 'gene', id = gene, qualifiers = qualifiers,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cd0a6d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant and other gene features\n",
    "def feature_key(feature):\n",
    "    # need to order the other features in a specific way\n",
    "    # this is the key function\n",
    "    if feature.type == 'gene':\n",
    "        return 0\n",
    "    if 'is_sig' in feature.qualifiers and feature.qualifiers['is_sig']:\n",
    "        return 3\n",
    "    return 1\n",
    "\n",
    "var_data = pd.read_csv(data_path.parent / 'variant_analysis.csv')\n",
    "other_genes = pd.read_csv(data_path / 'other_genes.csv')\n",
    "var_features_dict = {}\n",
    "for gene in list(plasmo_gene_data.name.unique()) + list(neis_gene_data.name.unique()):\n",
    "    \n",
    "    gene_vars = var_data.loc[var_data.gene == gene]\n",
    "    var_features = [\n",
    "        SeqFeature(\n",
    "            SimpleLocation(int(r['pos']), int(r['pos']) + 3\n",
    "        ), type = 'variant', qualifiers = {'is_sig': r['is_sig']})\n",
    "        for r in gene_vars.to_dict('records')\n",
    "    ]\n",
    "    og_slice = other_genes.loc[other_genes.parent == gene]\n",
    "    og_features = [\n",
    "        SeqFeature(SimpleLocation(\n",
    "            int(row['cds_start']), int(row['cds_end']),\n",
    "            strand = 1 if row['direction'] == 'forward' else -1,\n",
    "        ), type = 'gene', id = row['name'], )\n",
    "        for row in og_slice.to_dict('records')\n",
    "    ]\n",
    "    var_features_dict[gene] = sorted(og_features + var_features, key=feature_key, reverse=False)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f50d3ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_color(feature):\n",
    "    # defines the color map of features\n",
    "    if feature.id.endswith('utr'):\n",
    "        return 'white'\n",
    "    if feature.type == 'variant':\n",
    "        if feature.qualifiers['is_sig']:\n",
    "            return 'red'\n",
    "        return 'yellow'\n",
    "    if feature.strand == -1:\n",
    "        return 'orange'\n",
    "    return 'blue'\n",
    "\n",
    "diagrams = []\n",
    "for g in gene_features:\n",
    "    gd_diagram = GenomeDiagram.Diagram(\n",
    "        g.id, \n",
    "        start=g.qualifiers['segment'], \n",
    "        end=g.qualifiers['segment'] + g.qualifiers['segment_length'],\n",
    "        track_size = 0.6,\n",
    "    )\n",
    "    track = gd_diagram.new_track(\n",
    "        1, name=g.id, \n",
    "        scale=1, scale_format='SInt', scale_color=colors.black, greytrack = True\n",
    "    )\n",
    "    feature_set = track.new_set()\n",
    "    feature_set.add_feature(\n",
    "        g, sigil=\"BOX\", color = get_color(g), \n",
    "        name = g.id, label=False, label_size=28, label_angle=0 if g.strand == 1 else 180,\n",
    "        label_position = 'middle',\n",
    "    )\n",
    "    if g.id in var_features_dict:\n",
    "        for v in var_features_dict[g.id]:\n",
    "            feature_set.add_feature(\n",
    "                v, color = get_color(v), name = v.id,\n",
    "                sigil=\"BOX\" if v.type == 'gene' else \"BOX\", \n",
    "                label = False if v.type == 'gene' else False,\n",
    "                label_size=28, label_angle=0 if g.strand == 1 else 180,\n",
    "                label_position = 'middle'\n",
    "            )\n",
    "    diagrams.append(gd_diagram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "78a06f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# I had to turn off the labels because they were hard to deal with. I'll add them manually later\n",
    "for d, g in zip(diagrams, gene_features):\n",
    "    d.draw(format='linear', fragments = 1)\n",
    "    d.write(data_path.parent / 'var_svgs' / f\"{g.id}.svg\", \"SVG\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
