{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c153829a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "import re\n",
    "from scipy.stats import ttest_ind\n",
    "from Bio.Seq import Seq\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9addddac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data_path = Path('/mnt/d/data/popnet_paper')\n",
    "data_sets = ['purE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0cd2cfa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def infer_wt(col):\n",
    "    \"\"\"col is a column with the genotype calls of each sample at a given position.\"\"\"\n",
    "    # we don't count dots\n",
    "    counts = col.value_counts()\n",
    "    # returns WT, Mutant\n",
    "    # the results of value_counts are sorted in descending order by default\n",
    "    return counts.index[0], counts.index[1]\n",
    "\n",
    "def get_mutation(seq, pos, wt, mut, reverse=False):\n",
    "    # adjust for zero index\n",
    "    if reverse:\n",
    "        pos = pos + 1\n",
    "    else:\n",
    "        pos = pos - 1\n",
    "\n",
    "    if seq[pos] != wt:\n",
    "        raise ValueError(\"WT sequences don't match\")\n",
    "    mut_seq = seq[:pos] + mut + seq[pos + 1:]\n",
    "    offset = pos % 3\n",
    "    wp = Seq(seq[pos-offset:pos+3-offset]).translate()\n",
    "    mp = Seq(mut_seq[pos-offset:pos+3-offset]).translate()\n",
    "    return wp, mp \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51344b42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# perform one-way ANOVA\n",
    "results = []\n",
    "with open(data_path / 'neis_meta.tsv', 'r') as f:\n",
    "    meta = [int(float(x)) for x in f.read().splitlines()]\n",
    "for gene in data_sets:\n",
    "    df = pd.read_csv(data_path / f'{gene}.csv')\n",
    "    df['meta'] = meta\n",
    "    y_col = 'meta'\n",
    "    x_cols = [col for col in df.columns if re.match(r'^[0-9]+', col)]\n",
    "    for x_col in x_cols:\n",
    "        wt, mut = infer_wt(df[x_col]) # since there are many mutants, we just take the first one as WT\n",
    "        wt_sample = df.loc[df[x_col] == wt, y_col]\n",
    "        mut_sample = df.loc[df[x_col] != wt, y_col]\n",
    "        _, p_val = ttest_ind(wt_sample, mut_sample, equal_var=False, nan_policy='omit', alternative='two-sided')\n",
    "        results.append({\n",
    "            'gene': gene,\n",
    "            'position': x_col,\n",
    "            'p_val': p_val,\n",
    "            'wt': wt,\n",
    "            'mut': mut,\n",
    "        })\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df['p_adj'] = multipletests(results_df['p_val'].values)[1]\n",
    "filtered_results = results_df.loc[results_df['p_adj'] < 0.01]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f413278d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>position</th>\n",
       "      <th>p_val</th>\n",
       "      <th>wt</th>\n",
       "      <th>mut</th>\n",
       "      <th>p_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>purE</td>\n",
       "      <td>745887</td>\n",
       "      <td>0.000373</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.005205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>purE</td>\n",
       "      <td>745891</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>purE</td>\n",
       "      <td>745916</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene position     p_val wt mut     p_adj\n",
       "13  purE   745887  0.000373  A   G  0.005205\n",
       "14  purE   745891  0.000003  T   C  0.000051\n",
       "15  purE   745916  0.000003  A   C  0.000051"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83780c4e",
   "metadata": {},
   "source": [
    "All three significant variants are downstream of the coding reason, in the intergenic region between purE and transposase NGO_RS11705 (also downstream because it's anti-sense). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f1c87f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
