{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c153829a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from copy import deepcopy\n",
    "from pathlib import Path\n",
    "from scipy.stats import ttest_ind, fisher_exact\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "import numpy as np\n",
    "from itertools import chain\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0cd2cfa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we're actually going to drop the clearance time in file in favor of a centralized one\n",
    "def clean_var_data(var_data):\n",
    "    return var_data[['POS'] + [x for x in var_data.columns if x.isnumeric()]]\n",
    "\n",
    "# we need to summarize the variants data and do the one-way anova before trying anything else\n",
    "def process_var_data(variants, labels, ttest=True):\n",
    "    results = []\n",
    "    positions = [ x for x in variants.columns if x.isnumeric()]\n",
    "    for p in positions:\n",
    "        var_counts = variants[p].value_counts()\n",
    "        wt_names = [ x for x in variants.loc[variants[p] == var_counts.index[0], 'POS'].values if x in labels.index] # the pos column is actually the name column\n",
    "        if len(var_counts) > 1:\n",
    "            for idx in var_counts.index[1:]: # idx is the variant\n",
    "                var_names = [ x for x in variants.loc[variants[p] == idx, 'POS'].values if x in labels.index]\n",
    "                assert abs(len(var_names) - var_counts[idx]) <= 1 # can be off by one cuz we don't have some labels\n",
    "                if ttest:\n",
    "                    _, p_val = ttest_ind(\n",
    "                        [float(x) for x in labels.loc[wt_names] if not pd.isna(x)],\n",
    "                        [float(x) for x in labels.loc[var_names] if not pd.isna(x)],\n",
    "                        equal_var=False, nan_policy='omit', \n",
    "                        alternative='two-sided'\n",
    "                    )\n",
    "                else:\n",
    "                    #binary variables need chisquared\n",
    "                    wt_labels = [float(x) for x in labels.loc[wt_names] if not pd.isna(x)]\n",
    "                    var_labels = [float(x) for x in labels.loc[var_names] if not pd.isna(x)]\n",
    "                    contingency_table = np.array([\n",
    "                        [sum(wt_labels), (len(wt_labels) - sum(wt_labels))],\n",
    "                        [sum(var_labels), (len(var_labels) - sum(var_labels))],\n",
    "                    ])\n",
    "                    if np.min(contingency_table) > 0:\n",
    "                        p_val = fisher_exact(contingency_table)[1]\n",
    "                    else:\n",
    "                        p_val = 1 # if we don't have enough data to get a pval it's a pass\n",
    "                results.append({\n",
    "                    'pos': p,\n",
    "                    'wt': var_counts.index[0],\n",
    "                    'var': idx,\n",
    "                    'count': len(var_names),\n",
    "                    'p_val': p_val,\n",
    "                })\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "\n",
    "# gene_data is a slice of the dataframe\n",
    "def check_synonymous(variants, gene_data, sequence, prot_sequence):\n",
    "    # the reverse genes are on the complement trans and backwords, so we'll need\n",
    "    # to change the nucleotide as well\n",
    "    complement_map = {\n",
    "        'A': 'T',\n",
    "        'T': 'A',\n",
    "        'C': 'G',\n",
    "        'G': 'C',\n",
    "    }\n",
    "\n",
    "    if len(gene_data['direction'].unique()) != 1:\n",
    "        raise ValueError('Multiple directions')\n",
    "    \n",
    "    is_forward = gene_data['direction'].unique()[0] == 'forward'\n",
    "    # construct the nucleotide mapping\n",
    "    cds_positions = list(chain.from_iterable([\n",
    "        range(start, end + 1)\n",
    "        for start, end in \n",
    "        zip(gene_data.cds_start.values, gene_data.cds_end.values)\n",
    "    ]))\n",
    "\n",
    "    if 'reverse' in gene_data.direction.values:\n",
    "        cds_positions.reverse()\n",
    "    \n",
    "    # here we have genome position: \n",
    "    pos_map = {pos: idx for idx, pos in enumerate(cds_positions)}\n",
    "\n",
    "    syn_results = []\n",
    "    # now we can apply the variants\n",
    "    for r in variants.to_dict('records'):\n",
    "\n",
    "        # we don't consider variants not in the gene range for this function\n",
    "        position = int(r['pos'])\n",
    "        aa_pos = None\n",
    "        old_aa = None\n",
    "        new_aa = None\n",
    "        if position not in pos_map:\n",
    "            syn = None\n",
    "        elif len(r['var']) != 1 or len(r['wt']) != 1 or r['var'] == '.' or r['wt'] == '.':\n",
    "            syn = False  # if the variant is not single nuc, it's always non-syn\n",
    "        else:\n",
    "            rp = pos_map[position] # position to replace\n",
    "            wt = r['wt'] if is_forward else complement_map[r['wt']]\n",
    "            var = r['var'] if is_forward else complement_map[r['var']]\n",
    "            wt_seq = sequence[:rp] + wt + sequence[rp + 1:]\n",
    "            new_seq = sequence[:rp] + var + sequence[rp + 1:] # replace one character\n",
    "            wt_prot = str(Seq(wt_seq).translate())[:-1]\n",
    "            new_prot = str(Seq(new_seq).translate())[:-1] # the last char is the stop codon\n",
    "            try:\n",
    "                assert len(new_prot) == len(wt_prot) == len(prot_sequence)\n",
    "            except AssertionError:\n",
    "                print(r)\n",
    "                raise AssertionError\n",
    "            syn = (new_prot == wt_prot)\n",
    "            aa_pos = int(rp / 3) + 1\n",
    "            old_aa = wt_prot[aa_pos - 1]\n",
    "            new_aa = new_prot[aa_pos - 1]\n",
    "        \n",
    "        r['is_synom'] = syn\n",
    "        r['aa_pos'] = aa_pos\n",
    "        r['old_aa'] = old_aa\n",
    "        r['new_aa'] = new_aa\n",
    "        syn_results.append(r)\n",
    "    \n",
    "    return pd.DataFrame(syn_results)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9addddac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data_path = Path('/mnt/d/data/popnet_paper/variant_data')\n",
    "plasmo_gene_data = pd.read_csv(data_path / 'plasmo_genes.csv')\n",
    "neis_gene_data = pd.read_csv(data_path / 'neis_genes.csv')\n",
    "\n",
    "plasmo_genes = plasmo_gene_data['name'].unique()\n",
    "neis_genes = neis_gene_data['name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b87028c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# process plasmo genes\n",
    "plasmo_labels = pd.read_csv(data_path / 'plasmo_meta.tsv', sep='\\t', index_col = 0).rename({'Parasites clearance time': 'label'}, axis=1)\n",
    "results = []\n",
    "for gene in plasmo_genes:\n",
    "    var_data = clean_var_data(pd.read_csv(data_path / f\"{gene}.csv\"))\n",
    "    with open(data_path / f'{gene}_cds.fasta') as f:\n",
    "        sequence = [str(x.seq) for x in SeqIO.parse(f, 'fasta')][0] # there is always only one record per file\n",
    "    with open(data_path / f'{gene}_prot.fasta') as f:\n",
    "        prot_sequence = [str(x.seq) for x in SeqIO.parse(f, 'fasta')][0]\n",
    "    var_summary = process_var_data(var_data, plasmo_labels['label'])\n",
    "    gene_data_slice = plasmo_gene_data.loc[plasmo_gene_data['name'] == gene]\n",
    "    var_w_syn = check_synonymous(\n",
    "        var_summary, gene_data_slice, sequence, prot_sequence,\n",
    "    )\n",
    "    var_w_syn['chr'] = gene_data_slice['region'].apply(lambda x: x.split('_')[1]).unique()[0]\n",
    "    var_w_syn['gene'] = gene\n",
    "    var_w_syn['species'] = 'plasmodium'\n",
    "    results.append(var_w_syn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9b0320ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run analysis for neisseria\n",
    "labels = pd.read_csv(data_path / 'neis_meta.tsv', sep='\\t').set_index('Sample')\n",
    "gene_data = neis_gene_data\n",
    "for gene in neis_genes:\n",
    "    try:\n",
    "        var_data = clean_var_data(pd.read_csv(data_path / f\"{gene}.csv\"))\n",
    "    except FileNotFoundError:\n",
    "        continue\n",
    "\n",
    "    var_summary = process_var_data(var_data, labels['label'], ttest=False)\n",
    "    gene_data_slice = gene_data.loc[gene_data['name'] == gene]\n",
    "    with open(data_path / f'{gene}_cds.fasta') as f:\n",
    "        sequence = [str(x.seq) for x in SeqIO.parse(f, 'fasta')][0] # there is always only one record per file\n",
    "    with open(data_path / f'{gene}_prot.fasta') as f:\n",
    "        prot_sequence = [str(x.seq) for x in SeqIO.parse(f, 'fasta')][0]\n",
    "\n",
    "    var_w_syn = check_synonymous(\n",
    "        var_summary, gene_data_slice, sequence, prot_sequence,\n",
    "    )\n",
    "\n",
    "    var_w_syn['chr'] = 1\n",
    "    var_w_syn['gene'] = gene\n",
    "    var_w_syn['species'] = 'neisseria'\n",
    "    results.append(var_w_syn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "459d9c81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/javi/workspace/NeuralNet/venv/lib/python3.7/site-packages/statsmodels/stats/multitest.py:160: RuntimeWarning: invalid value encountered in greater\n",
      "  notreject = pvals > alphacSidak_all\n",
      "/home/javi/workspace/NeuralNet/venv/lib/python3.7/site-packages/statsmodels/stats/multitest.py:177: RuntimeWarning: divide by zero encountered in log1p\n",
      "  np.log1p(-pvals))\n",
      "/home/javi/workspace/NeuralNet/venv/lib/python3.7/site-packages/statsmodels/stats/multitest.py:257: RuntimeWarning: invalid value encountered in greater\n",
      "  pvals_corrected[pvals_corrected>1] = 1\n"
     ]
    }
   ],
   "source": [
    "results_df = pd.concat(results)\n",
    "results_df['p_adj'] = multipletests(results_df['p_val'].values)[1]\n",
    "results_df['is_sig'] = results_df['p_adj'] < 0.01\n",
    "\n",
    "results_df = results_df[[\n",
    "    'species', 'gene', 'chr', 'pos', 'wt', 'var', 'p_adj', 'is_sig', 'count', 'aa_pos', 'old_aa', 'new_aa', 'is_synom',\n",
    "]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "708e851f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>segment_name</th>\n",
       "      <th>total_var</th>\n",
       "      <th>sig_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>neisseria</td>\n",
       "      <td>NgS1</td>\n",
       "      <td>32</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>neisseria</td>\n",
       "      <td>NgS2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>neisseria</td>\n",
       "      <td>NgS3</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>neisseria</td>\n",
       "      <td>NgS4</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>plasmodium</td>\n",
       "      <td>PfS1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>plasmodium</td>\n",
       "      <td>PfS2</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>plasmodium</td>\n",
       "      <td>PfS3</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>plasmodium</td>\n",
       "      <td>PfS4</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>plasmodium</td>\n",
       "      <td>PfS5</td>\n",
       "      <td>22</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      species segment_name  total_var  sig_count\n",
       "0   neisseria         NgS1         32          3\n",
       "1   neisseria         NgS2          4          1\n",
       "2   neisseria         NgS3         10          0\n",
       "3   neisseria         NgS4         46          0\n",
       "4  plasmodium         PfS1         12          1\n",
       "5  plasmodium         PfS2         16          1\n",
       "6  plasmodium         PfS3         12          5\n",
       "7  plasmodium         PfS4         14          5\n",
       "8  plasmodium         PfS5         22          4"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# construct the per-segment table\n",
    "gene_seg_map = {\n",
    "    'med14': 'PfS1',\n",
    "    'pf': 'PfS2',\n",
    "    'pmt': 'PfS3',\n",
    "    'kelch': 'PfS4',\n",
    "    '6pgd': 'PfS5',\n",
    "    'pure': 'NgS1',\n",
    "    'barrel': 'NgS2',\n",
    "    'is110': 'NgS3',\n",
    "    'c39': 'NgS4',\n",
    "}\n",
    "results_df['segment_name'] = results_df.gene.map(gene_seg_map)\n",
    "segment_summary = results_df.groupby(['species', 'segment_name']).agg(\n",
    "    total_var = pd.NamedAgg(column='is_sig', aggfunc=\"count\"),\n",
    "    sig_count = pd.NamedAgg(column='is_sig', aggfunc=\"sum\"),\n",
    ").reset_index(drop=False)\n",
    "segment_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fbc19dfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "segment_summary.to_csv(data_path.parent / 'segment_summary.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
